### PR TITLE
[GEOT-7585] Css unique rule names

### PR DESCRIPTION
**Pull request by sweco-se1c8x**
[GEOT-7585](https://osgeo-org.atlassian.net/browse/GEOT-7585) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

The primary objective of this Pull Request is to introduce the UniqueRuleNames directive into the GeoServer CSS styling system. It provides Geoserver with the ability to include rule names when querying a layer employing a style created with the CSS extension for the json representation of its legend:
`wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetLegendGraphic&FORMAT=application/json&LAYER=topp:tasmania_roads&SCALE=100`
 
(just like for styles written as SLD that include rule names)
 
The individual rule names received can then be employed to fetch rule specific legend graphics:
wms?REQUEST=GetLegendGraphic&VERSION=1.3.0&FORMAT=image/png&LAYER=topp:tasmania_roads&style=tasmania_roads&RULE=0
 
Which is sometimes more useful than the getlegendgraphic complete-legend composite image for theme layers.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [ ] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [ ] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Review by aaime — CHANGES_REQUESTED** [review]
The change is small but the objective is unclear, the description seems to have been written by a marketing department.

So let's have a look at what it does... after generating the Style object, it goes over every rule, in whatever order it has been generated, and adds a progressive number to them, starting from zero and going up.

In the general case, where SLD rules are generated by cascading the CSS rules, I have a hard time understanding how one would be able to predict which rule will be given which number, thus "improve the usability and interpretability of map legends". A flat list of numbers does nothing for "interpretability", although I agree it makes it possible to "allowing individual symbol retrieval in a thematic style".

Thinking out loud, it seems it would still help in when "flat" mode is enabled, so that no cascading is actually used. Maybe the target use case is this one, but it should be clarified.

The Jira ticket shows more marketing department speech, which should be simplified and made to match the actual PR behavior, e.g.:

* "This issue leads to confusion among users, as the legend fails to reflect the actual symbology used on the map"
* " The integration of the uniqueRuleNames directive is crucial for the legend generation process to support the aggregation of multiple rules into a single legend item, thereby providing a coherent and accurate representation of the map's symbology."
* "By doing so, we aim to enable the generation of legends that can capture the full visual complexity of compound symbols, such as those used for highways, and improve the overall interpretability of map legends for end-users."

All the PR does it to allow pinpoint a particular rule in GetLegendGraphics, but I fail to see how that helps with the above. It must be related to the specific way you generated and use legends in GeoServer, which should be clarified.

Finally, no PR is merged without a corresponding test, please add one

**Comment by steff-o** [comment]
Sorry for barging in, but I think this is a great addition and I will try to explain why I also would like to have this feature.

This is useful when creating nice looking legends in an application where it is possible to toggle visibility on individual symbols. There is an open source project, Origo, https://github.com/origo-map/origo, maintained by several municipalities in Sweden that uses this way of fetching symbols by rule name, but it only works for SLD, as CSS styling has no name on the rules.

The basic flow is as follows:
1. The application fetches the symbology as json to determine which symbols are available
2. The application examines the json document and fetctes the symbol for each rule by rule name and creates a legend displaying all symbols with an accompanying radio button for toggling that particular symbol.
3. When a symbol is toggled, the associated filter from the json response is applied to the layer source.
4. The application also has a high resolution print view, which fetches all maps in higher dpi in order to avoid small symbology when printing in high res. The getLegendGrpaphics does not have a DPI parameter, but by fetching a larger size and shrink it into a smaller container it will produce a nice looking high res print. By fetching each symbol individually produces a nicer looking high res print than just fetching the entire legendgraphic as the texts can be handled separately. As a bonus in that view, only rules that are selected visible in the legend are displayed on the print to save space. 

Given the functionality above, the actual name of the rule is not important, it just needs to be unique.

This is all working and widely used in production, but only for SLD. It would be useful if it worked for CSS as well, as our map authors find CSS easier to work with. As now, they create the rules in CSS, converts to SLD and names all rules, which is tedious and error prone.

**Comment by aaime** [comment]
@steff-o that's exactly the explanation that I want to see in the documentation, as an example of why the flag is useful and how it's intended to be used, how to make sense of the apparently random numbers attached to rules as names.

**Comment by sweco-se1c8x** [comment]
@aaime The variable name has been updated to autoRoleNames, and the documentation has been revised based on steff-o description. Could you please provide an update on the status of the pull request? And if there are any more changes needed?

**Comment by aaime** [comment]
This week I'm stuck with a few work emergencies and I could not perform the usual Monday PR review round. Next week, hopefully.

**Comment by aaime** [comment]
All the github actions failed... to my surprise, they are all legit failures. Did you try to compile the code before submitting it?

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:testCompile (default-testCompile) on project gt-geopkg: Compilation failure: Compilation failure: 
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[19,28] cannot find symbol
Error:    symbol:   class FeatureTypeStyle
Error:    location: package org.geotools.styling
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[20,28] cannot find symbol
Error:    symbol:   class Rule
Error:    location: package org.geotools.styling
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[21,28] cannot find symbol
Error:    symbol:   class Style
Error:    location: package org.geotools.styling
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[22,32] package org.geotools.styling.css does not exist
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[28,13] cannot find symbol
Error:    symbol:   class Stylesheet
Error:    location: class CssTranslatorTest
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[33,26] cannot find symbol
Error:    symbol:   class Stylesheet
Error:    location: class CssTranslatorTest
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[41,33] cannot find symbol
Error:    symbol:   variable CssTranslator
Error:    location: class CssTranslatorTest
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[43,9] cannot find symbol
Error:    symbol:   class Style
Error:    location: class CssTranslatorTest
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[45,14] cannot find symbol
Error:    symbol:   class FeatureTypeStyle
Error:    location: class CssTranslatorTest
Error:  /home/runner/work/geotools/geotools/modules/plugin/geopkg/src/test/java/org/geotools/geopkg/CssTranslatorTest.java:[46,18] cannot find symbol
Error:    symbol:   class Rule
Error:    location: class CssTranslatorTest
```

Why did you add a test class in the "org.geotools.geopkg" package? It cannot see the classes it needs, because it's in the wrong package. Also, while you're at it, please just add one new test method in the existing "TranslatorSyntheticTest" instead, and then build with maven, with a java 11, something like this:

```
cd modules\unsupported\css
mvn clean install -Dqa
```

That will allow you to verify the module builds, that it passes the QA checks, and also ensure the sources are formatted according to convention (just let it run, it will reformat the code to match the expected output).

**Comment by sweco-se1c8x** [comment]
Hello @aaime! I encounter a certification error when running the test with -Dqa. Conversely, running it without this option results in a different error, as shown in the image below:

image
I am unsure how to fix this issue since the message doesn't provide much clarity. Could you please assist me in resolving this error?

**Comment by aaime** [comment]
That's a generic network error happening during the maven build, it has nothing to do with the QA flags.
I'd recommend rebuilding with the -U flag, it will force maven to re-attempt downloading missing dependencies from the maven repositories.

**Comment by sweco-se1c8x** [comment]
Thank you for your answer, I tried with -U but still the same error. Here is the command I am running
`mvn clean install -U -Dmaven.wagon.http.ssl.insecure=true -Dmaven.wagon.http.ssl.allowall=true -Dmaven.wagon.http.ssl.ignore.validity.dates=true`

Running it without -Dmaven.wagon.http.ssl.insecure=true -Dmaven.wagon.http.ssl.allowall=true -Dmaven.wagon.http.ssl.ignore.validity.dates=true i am getting certifcation error.

**Comment by aaime** [comment]
Looks like you're running maven in a network with a broken proxy. Well, it's the last module in the build, if the rest of it works fine, just go ahead and push the changes, let's see what the Github Actions make of it.

**Comment by sweco-se1c8x** [comment]
Ok good, i will push the changes now. Thanks

**Comment by aaime** [comment]
@sweco-se1c8x it seems you're trying to use the Github actions as a substitute for a local compiler. 
Could you do that locally instead? 

Since all your changes are in one module, take a recent Maven, a Java11 or newer, a GeoTools checkout, and do the following until the build succeeds:

```
cd modules/unsupported/css/
mvn clean install -Dqa
```

That builds just the CSS module, with QA enabled.

Then to make sure everything is aligned, get a GeoServer checkout and do:

```
cd src/extensions/css
mvn clean install -Dqa
```

When both builds are working you're ready to make a "git push".

**Comment by sweco-se1c8x** [comment]
@aaime, I understand it's not ideal to run all those commits, but the tests aren't working locally. I have no clue why, and the error messages aren't very informative. But in the other hand i am getting the error in the github actions. 

`[ERROR] Errors:
[ERROR]   TranslatorSyntheticTest.testTranslateWithAutoNames:140 » IllegalArgument Inval...
[INFO]
[ERROR] Tests run: 346, Failures: 0, Errors: 1, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  48.879 s
[INFO] Finished at: 2024-06-24T12:33:10+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.1:test (default-test) on project gt-css: There are test failures.
[ERROR]
[ERROR] Please refer to C:\Users\se1c8x\project\geotools\modules\unsupported\css\target\surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException`

**Comment by aaime** [comment]
Hum... the current version of the code is not even compiling, and the code formatted breaks on it:

```
[INFO] --- spotless:2.20.0:apply (default) @ gt-css ---
[INFO] Up-to-date checking enabled
giu 24, 2024 12:55:01 PM com.diffplug.spotless.FormatExceptionPolicyLegacy error
GRAVE: Step 'google-java-format' found problem in 'src/test/java/org/geotools/styling/css/TranslatorSyntheticTest.java':
You are not using latest version on JVM 11+.
Try to upgrade to google-java-format 1.12.0, which may have fixed this problem.
java.lang.Exception: You are not using latest version on JVM 11+.
Try to upgrade to google-java-format 1.12.0, which may have fixed this problem.
	at com.diffplug.spotless.Jvm$Support$1.apply(Jvm.java:181)
	at com.diffplug.spotless.FormatterStepImpl$Standard.format(FormatterStepImpl.java:78)
	at com.diffplug.spotless.FormatterStep$Strict.format(FormatterStep.java:88)
	at com.diffplug.spotless.Formatter.compute(Formatter.java:230)
	at com.diffplug.spotless.PaddedCell.calculateDirtyState(PaddedCell.java:203)
	at com.diffplug.spotless.PaddedCell.calculateDirtyState(PaddedCell.java:190)
	at com.diffplug.spotless.maven.SpotlessApplyMojo.process(SpotlessApplyMojo.java:45)
	at com.diffplug.spotless.maven.AbstractSpotlessMojo.execute(AbstractSpotlessMojo.java:191)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:126)
	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2(MojoExecutor.java:328)
	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:316)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:174)
	at org.apache.maven.lifecycle.internal.MojoExecutor.access$000(MojoExecutor.java:75)
	at org.apache.maven.lifecycle.internal.MojoExecutor$1.run(MojoExecutor.java:162)
	at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute(DefaultMojosExecutionStrategy.java:39)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:159)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:105)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:73)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:53)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:118)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:261)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:173)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:101)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:906)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:283)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:206)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:283)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:226)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:407)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:348)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.diffplug.spotless.java.GoogleJavaFormatStep$State.lambda$createFormat$1(GoogleJavaFormatStep.java:176)
	at com.diffplug.spotless.FormatterFunc.apply(FormatterFunc.java:32)
	at com.diffplug.spotless.Jvm$Support$1.apply(Jvm.java:179)
	... 34 more
Caused by: com.google.googlejavaformat.java.FormatterException: 140:66: error: ';' expected
	at com.google.googlejavaformat.java.FormatterException.fromJavacDiagnostics(FormatterException.java:50)
	at com.google.googlejavaformat.java.Formatter.format(Formatter.java:151)
	at com.google.googlejavaformat.java.Formatter.getFormatReplacements(Formatter.java:258)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:234)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:202)
	... 41 more
```

One can skip the formatted to receive the error message in a more succint way:

```
mvn clean install -nsu -fae -Dspotless.apply.skip=true
...
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:testCompile (default-testCompile) on project gt-css: Compilation failure
[ERROR] /home/aaime/devel/git-gt/modules/unsupported/css/src/test/java/org/geotools/styling/css/TranslatorSyntheticTest.java:[140,65] ';' expected
```

**Comment by aaime** [comment]
The test failure seem clear to me:

```
2024-06-24T12:49:25.4242850Z [ERROR] Tests run: 106, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.147 s <<< FAILURE! - in org.geotools.styling.css.TranslatorSyntheticTest
2024-06-24T12:49:25.4244980Z [ERROR] testTranslateWithAutoNames(org.geotools.styling.css.TranslatorSyntheticTest)  Time elapsed: 0.005 s  <<< ERROR!
2024-06-24T12:49:25.4247316Z java.lang.IllegalArgumentException: Invalid CSS style, no rule seems to activate any symbolization. The properties activating the symbolizers are fill, stroke, mark, label, raster-channels, have any been used in a rule matching any feature?
```

So the style you're trying to translate is not a valid CSS style, it lacks the minimum amount of properties (in particular, one of fill, stroke, mark, label, raster-channels)

**Comment by sweco-se1c8x** [comment]
I was creating an empty StyleSheet, which is why I encountered so many errors. Now I've seen the CssTestBase class, and I've started using it instead. However, I'm still getting an error, the same one that I'm getting locally (Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and ). @aaime, do you have any idea why I might be receiving this error?
I did a few searches and found that others are experiencing the same error, but there's no direct answer on how to resolve it.

File: modules/unsupported/css/src/test/java/org/geotools/styling/css/TranslatorSyntheticTest.java [hunk 1]
```diff
@@ -127,6 +127,26 @@ public void fillBasic() throws IOException {
         assertNull(fill.getGraphicFill());
     }
 
+    @Test
+    public void testTranslateWithAutoNames() {
+        // Define the CSS string
+        String css = "@autoRuleNames 'true';\n" + "* { fill: #FF0000; }";
```
Comment (aaime) [inline modules/unsupported/css/src/test/java/org/geotools/styling/css/TranslatorSyntheticTest.java:133]: I've changed the code to actually test what a user would write (the actual directive syntax). 
If the builds are green, I'll merge.

**Comment by aaime** [comment]
Squash-merged down to one commit.

**Pull request merged**